### PR TITLE
Call remote authors

### PR DESCRIPTION
**Pull request by TurnipXenon**
# Overview

Make local node make remote calls on your behalf.

## Design decision

To indicate we want to do things outside our local node. We add a node parameter, which should be equal to the domain of that server.

This is based on how Mastodon does it (partially) where they append the url with the remote server the account was originally from, like `https://mastodon.yourserver/@account@differentserver`

Ours follow: `http://service/endpoint/?node=domain.herokuapp.com`

## Changes
- Add remote logic for Author endpoint
- Add RemoteConfigs which has overrideable values and logic that can be on a per server (per python class) basis

## Testing
Go to the following urls
- See all local authors: https://potato-oomfie.herokuapp.com/authors/
- See all remote authors: https://potato-oomfie.herokuapp.com/authors/?node=turnip-oomfie-1.herokuapp.com
- See a local author: http://turnip-oomfie-1.herokuapp.com/authors/656821bc-6ca1-4fe3-ba1b-927e9860d0a4/
- See a remote author: http://potato-oomfie.herokuapp.com/authors/656821bc-6ca1-4fe3-ba1b-927e9860d0a4/?node=turnip-oomfie-1.herokuapp.com
- Pagination: https://turnip-oomfie-1.herokuapp.com/authors/?node=potato-oomfie.herokuapp.com&page=2&size=1

### CODE REVIEW

File: mysocial/authors/models/author.py [hunk 1]
```diff
@@ -93,3 +100,27 @@ def get_all_authors():
...
+
+    # check if we have this server
+    if domain not in RemoteUtil.CONFIG:
+        raise ValidationError(f'{author_url} does not have any corresponding domain')
+
+    # todo: otherwise, check it at the other server
+    node_config = RemoteUtil.CONFIG[domain]
+    author = node_config.get_author(author_url)
+    if author is None:
+        raise ValidationError(f'{author_url} does not exist in the domain {domain}')
```
Comment (TurnipXenon) [inline mysocial/authors/models/author.py:126]: Ignore this one for now. It doesn't make sense (YET).

File: mysocial/follow/models.py [hunk 2]
```diff
@@ -1,17 +1,18 @@
...
     """
     FIELD_NAME_HAS_ACCEPTED = 'hasAccepted'
 
-    actor = models.ForeignKey(settings.base.AUTH_USER_MODEL, on_delete=models.CASCADE, related_name='+')
-    target = models.ForeignKey(settings.base.AUTH_USER_MODEL, on_delete=models.CASCADE, related_name='+')
+    actor = models.URLField(settings.base.AUTH_USER_MODEL, validators=[validate_author_url])
+    target = models.URLField(settings.base.AUTH_USER_MODEL, validators=[validate_author_url])
     has_accepted = models.BooleanField(default=False)
 
     class Meta:
```
Comment (TurnipXenon) [inline mysocial/follow/models.py:18]: Ignore for now. This one is for Follow remote.

File: mysocial/remote_nodes/node_config_base.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    def get_author(self, author_url: str):
+        token = base64.b64encode(f'{self.username}:{self.password}'.encode('ascii')).decode('utf-8')
+        headers = {'HTTP_AUTHORIZATION': f'Basic {token}'}
+        response = requests.get(author_url, **headers)
+
+        if response.status_code == 200:
+            # todo(turnip): map to our author?
+            return response.data
+        else:
+            return None
```
Comment (TurnipXenon) [inline mysocial/remote_nodes/node_config_base.py:69]: for follow remote change; ignore

**Comment by amandanguyeen** [comment]
You might need to clarify something for me. 
My assumptions: we have authors on our server1 and there are authors on server2. That author2 will want to connect to our api and so they register as an unauthorized user until they get approved by server admin. When they do get approved, we take their host an add it to the list of servers we accept. Then we have 2 GET methods: one to retrieve hosts on our server and one to retrieve hosts on another server is that correct?

Coordinating user story:
GET [local, remote]: retrieve all profiles on the server (paginated)

**Comment by TurnipXenon** [comment]
> Then we have 2 GET methods: one to retrieve hosts on our server and one to retrieve hosts on another server is that correct?
Is host here meaning a real Author? Or a Node?

I just made these 2 methods as one endpoint. We call this endpoint normally "to retrieve hosts on our server". We add a parameter `?node=another.server` "to retrieve hosts on another server".

> That author2 will want to connect to our api and so they register as an unauthorized user until they get approved by server admin. When they do get approved, we take their host an add it to the list of servers we accept.
Hmm... I kinda don't understand this one. Is Author2 an author in Node2? Is Author2 calling our Node (Node1)?

How I assumed it works is that, as an Author, I want my Node to make calls on my behalf to the remote Node.

It's based on this [Scenario Summary](https://github.com/abramhindle/CMPUT404-project-socialdistribution/blob/master/project.org#scenario-summary):

> All actions from authors are routed through the inbox of the receiving authors by the node itself.

**Comment by amandanguyeen** [comment]
>> I just made these 2 methods as one endpoint. We call this endpoint normally "to retrieve hosts on our server". We add a parameter `?node=another.server` "to retrieve hosts on another server". Does that design sound fine?
 ah okay, yeah i understand what the endpoints do my question is more, do we need them? 
>
> > That author2 will want to connect to our api and so they register as an unauthorized user until they get approved by server admin. When they do get approved, we take their host an add it to the list of servers we accept.
> > Hmm... I kinda don't understand this one. Is Author2 an author in Node2? Is Author2 calling our Node (Node1)?
> 
>
> How I assumed it works is that, as an Author, I want my Node to make calls on my behalf of me to the remote Node.
> 
> It's based on this [Scenario Summary](https://github.com/abramhindle/CMPUT404-project-socialdistribution/blob/master/project.org#scenario-summary):
> 
> > All actions from authors are routed through the inbox of the receiving authors by the **node itself**.
> 
> The key word here is node itself, and not the client or author. So, we do the operation on our node. Sorry, I'm kinda confused to, and I still don't have the whole picture.

**Comment by TurnipXenon** [comment]
I'm temporarily re-opening this one because my remote follow requests are dependent on a lot of changes here.

**Comment by amandanguyeen** [comment]
This is how you fetch a specific author:
http://potato-oomfie.herokuapp.com/authors/656821bc-6ca1-4fe3-ba1b-927e9860d0a4/?node=turnip-oomfie-1.herokuapp.com

to double check, the author id should be the author id of someone from the turbip oomfie server right?

**Review by amandanguyeen — APPROVED** [review]

**Pull request merged**